### DISCUSSION TITLE
GitHub Classroom abandoning its template-based approach breaks our workflow

### DISCUSSION THREAD

**Discussion by pcantrell** [community/community#78734] [GitHub Education] [answered]
I’m a college instructor. A few questions about the changes in this post:

https://github.blog/changelog/2023-12-08-upcoming-changes-to-assignment-repositories-and-starter-code-in-github-classroom/

- The easy question: Which org will student forks of assignments end up in? Will student forks still live the class’s org? Or will they now be scattered across the various student accounts?

- The hard question: Each one of our courses that uses GitHub Classroom currently has a **shared org** for assignments (like [this one](https://github.com/mac-comp127)), containing all of our starter code _and assignment solutions_. Students can’t see the solutions, because those repos are private; the assignment template repos are public.
 
  We reuse that shared org across semesters for assignment templates, so that corrections and updates to assignments go in the shared repo, and are carried forward to future semesters. Accepted student assignments, on the other hand, go in a **semester-specific org** (which is crucial to prevent the shared org from filling with hundreds upon hundreds of repos).

  However, under this new system GitHub is pushing out, for student repos to remain private from each other, the assignment starter repos must _also_ be private. As far as I can tell, these are our choices for handling that:
  
    1. Copy the starter code repo for _every assignment, every semester_ into the semester-specific org, a headache that adds a hurdle for making upstream fixes to assignments.
    1. Create a student group in the shared org, give that group read-only access to the now-private assignment repos, and add _every student every semester_ to that group.
    1. Give all the students full access to the shared org so they can see the private repos, exposing the solutions (and defeating the purpose of repos being private).

  Is there a non-headache-inducing option here I’m missing?

**Comment by RyanHecht** [comment, marked as answer]
Let me start with the easy question while I digest the hard one: 

> Which org will student forks of assignments end up in? Will student forks still live the class’s org? Or will they now be scattered across the various student accounts?

Student forks will all still live in the Classroom organization! GitHub recently shipped a feature allowing multiple forks of the same repository in the same namespace, which enabled us to make this change!

**Comment by RyanHecht** [comment]
As for your second question: Yes, we are adding an additional step to this workflow. I intend folks who make use of public starter code template repositories who still want their student repositories to be private to go with option (i):

> Copy the starter code repo for every assignment, every semester into the semester-specific org

You're absolutely right to note that this adds a hurdle for making upstream fixes to assignments, but let me point out: Because until now student repositories haven't been forks of the starter code repository, there's never been a way to get those fixes downstream to the already-accepted student repositories -- student repos are always a snapshot of the starter code at the time the assignment was accepted. 

Under the new system, should you discover a change to the starter code that has to be made mid-term after students have already accepted the assignment, you can do the following:
- Make the necessary change to the starter code, either the public copy in the shared org or the private copy in the semester-specific org
- [Add the other repository as a git remote on your client machine](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes#:~:text=Adding%20Remote%20Repositories) and push it/PR it into `main`
- Invite your students to use the GitHub GUI to [sync their fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) with the private starter code repository, something that was previously not possible when student repositories were made with templates

We understand the pain of this additional step for using public starter code repositories, but we're excited for the new use cases this enables with allowing teachers to push out changes to the starter code after the assignment has gone live -- something not possible before!

**Comment by pcantrell** [comment]
Thanks for the thorough answer, Ryan. Good to know about multiple forks sharing a namespace! That’s good news.

We are well aware of this limitation, and have been working around it quite happily:

> Because until now student repositories haven't been forks of the starter code repository, there's never been a way to get those fixes downstream to the already-accepted student repositories

Usually post-assignment upstream changes are small patches, and having students manually apply them is easy enough. Having to copy all the repos into the new org is a real headache. I’m sorry to report that, for us, at least, this solution will be substantially more painful than the problem it solves.

There was talk at one point of having some way of declaratively describing all the assignments for the course: a JSON file or some such containing all the GitHub Classroom settings for all the assignments in one big, reproducible batch, instead of having to work through the GUI to apply exactly the same settings for every assignment every semester. Did that ever launch? Such a file could also specify an upstream repo to copy into the classroom org, which would make the forking model much more practical.

**Reply by bmuellerhstat** [reply]
Echoing what @pcantrell said:

> this solution will be substantially more painful than the problem it solves.

Well said. I am in the exact same boat as you, with the exact same workflow. This forces me to make so many extra steps without a need for the new feature of pushing changes mid-assignment. Any future workarounds to this headache would be appreciated!

**Reply by RyanHecht** [reply]
Hey folks! Letting both @pcantrell and @bmuellerhstat know that the public beta for our transition to forks has begun: https://github.blog/changelog/2024-01-22-migrating-github-classroom-assignment-repository-creation-from-create-from-template-to-forks-public-beta/ 

Also letting you know that during the public beta period, our team is exploring options to automate some of the extra steps that are required with this new workflow. I really appreciate your feedback! :)
